074-k8s-storage/proposal.md

```diff
@@ -0,0 +1,172 @@
+	- Simplified GC
+	- Would support streaming across worker runtimes
+	
+### Cons
+- Some registries such as GCR don't expose an API to delete layers directly
+- **cache object** would have to have a fixed static path in the image file system to be able to reuse the same layer. This would require some additional handling on Concourse to support [input-mapping](https://concourse-ci.org/jobs.html#schema.step.task-step.input_mapping) and [output-mapping](https://concourse-ci.org/jobs.html#schema.step.task-step.output_mapping)
+- Adds extra development overhead to generate new image config & manifests to leverage **cache object** layers
+- Adds extra initialization overhead. Concourse wouldn't have control over the local caches on K8s nodes, so volumes would always have to be pushed to the centralized registry and pulled at least once when executing a step
+- Potentially adds substantial load on registry, as Concourse would be creating a new file system layer for every **cache object**
+- There isn't a well documented approach to setup an in-cluster secure registry. The setup requires exposing an in-cluster registry externally with traffic routed via an LB. [Prior spike](https://github.com/concourse/concourse/issues/3796)
+
+## S3 Compatible Blobstore
+## Description
+Each **cache object** is stored in a blobstore. Concourse would require a mananaged blobstore as a dependency. For each step, Concourse would pull down the relevant blobs for inputs and push blobs for outputs. 
+
+### Pros
+- Scale well (GCR uses GCS as the underlying storage)
+- Could explore further de-coupling by exploring CSI driver
+- Established ways of securely fetching/pushing blobs from an a blobstore
+- As this would be a centralized storage solution
+	- it doesn't impact what a K8s based Concourse worker looks like
+	- Simplified GC
+	- Would support streaming across worker runtimes
+	
+### Cons
+- Wouldn't have support for hosting images by default.
+- Adds another dependency for Concourse (depending on where Concourse is deployed there might be managed solutions available) 
+- Lack of standardized APIs
+- Adds extra initialization overhead. Concourse wouldn't have a local cache, so volumes would always have to be pushed & pulled for steps
+- Concourse would potentially be heavy user of the blobstore
+
+## Persistent Volumes
+Each **cache object** would be stored in its own persistent volume. Persistent volume snapshots would be used to reference **cache object** versions.
+
+### Pros
+- Would leverage native k8s offering
+- Maps well to Concourse's use of **cache objects** and offloads the heavy lifting to K8s
+- Potentially wouldn't require volumes to be streamed at all
+
+### Cons
+- Wouldn't have support for hosting images by default.
+- IaaS based limits on [volume limits per node](https://kubernetes.io/docs/concepts/storage/storage-limits/#dynamic-volume-limits) prevents this from being a scalable solution
```

Comment:
not super related, but found this idea interesting: https://github.com/argoproj/argo/issues/4130